src/commands/signzone.rs

```diff
@@ -0,0 +1,709 @@
+            use_nsec3,
+            algorithm,
+            iterations,
+            salt,
+            nsec3_opt_out_flags_only: true,
+            nsec3_opt_out,
+            hash_only: false,
+            zonefile_path,
+            key_paths,
+        })
+    }
+}
+
+impl SignZone {
+    pub fn parse_timestamp(arg: &str) -> Result<Timestamp, Error> {
+        // We can't just use Timestamp::from_str from the domain crate because
+        // ldns-signzone treats YYYYMMDD as a special case and domain does
+        // not. For invalid values this YYYYMMDDD prevents use of valid Unix
+        // timestamps that have the same value, e.g. ldns-signzone complains
+        // that for 99999999 "The month must be in the range 1 to 12". There's
+        // also no checking that an expiration timestamp is in the future of
+        // an inception timestamp (which for serial numbers is hard to say for
+        // sure but for YYYYMMDD or YYYYMMDDHHmmSS we could check).
+        let res = if arg.len() == 8 && arg.parse::<u32>().is_ok() {
+            // This can give strange errors, e.g. 99999999 warns about illegal
+            // signature time, but the alternative would be to add a
+            // dependency on chrono and parse the value ourselves in order to
+            // produce a better error message. Given that this only happens
+            // for very old or far future Unix timestamps we don't attempt to
+            // do better than this for now.
+            Timestamp::from_str(&format!("{arg}000000"))
+        } else {
+            Timestamp::from_str(arg)
+        };
+
+        res.map_err(|err| Error::from(format!("Invalid timestamp: {err}")))
+    }
+
+    pub fn execute(self, env: impl Env) -> Result<(), Error> {
+        // Post-process arguments.
+        // TODO: Can Clap do this for us?
+        let opt_out = if self.nsec3_opt_out {
+            Nsec3OptOut::OptOut
+        } else if self.nsec3_opt_out_flags_only {
+            Nsec3OptOut::OptOutFlagsOnly
+        } else {
+            Nsec3OptOut::NoOptOut
+        };
+
+        let signing_mode = if self.hash_only {
+            SigningMode::HashOnly
+        } else {
+            SigningMode::HashAndSign
+        };
+
+        let out_file = if let Some(out_file) = &self.out_file {
+            out_file.clone()
+        } else {
+            let out_file = format!("{}.signed", self.zonefile_path.display());
+            PathBuf::from_str(&out_file)
+                .map_err(|err| format!("Cannot write to {out_file}: {err}"))?
+        };
+
+        let mut writer = if out_file.as_os_str() == "-" {
+            // Box::new(env.stdout()) as Box<dyn Write>
+            // FIXME: env.stdout() uses impl fmt::Write, but because of
+            // domain::sign::records::SortedRecords::write_with_comments()
+            // we need io::Write here.
+            todo!()
+        } else {
+            Box::new(File::create(env.in_cwd(&out_file))?) as Box<dyn Write>
+        };
```

Comment:
Just so that it is written down: The merge of the env things introduced an issue in the current signzone code. Further down at L467 it uses writer with
```rust
records.write_with_comments(&mut writer, move |r, writer| match r.data() { ...
```
where it needs to implement `io::Write` currently, but `Env::stdout` uses `fmt::Write`. 
The idea mentioned earlier is to change the functions in `domain` to also use `fmt::Write`, as we don't need to be able to write arbitrary (non-UTF8) bytes.

Reply:
I updated the domain branch used by this PR to use `fmt::Write` instead of `io::Write`. The code in domain is in `src/sign/records.rs` which was in released code but never compiled as it had no `mod records.rs;` statement in `src/sign/mod.rs`. There is a `src/sign/bin/signzone.rs` orphaned example that might be affected so we should still check that. I have NOT updated this PR to use the new code in the domain branch.

Reply:
By the way, it might make more sense to:
- In the short term, move those functions out of domain into dnst.
- In the long term, add support to domain for some way of pretty printing zone entries with comments.
@tertsdiepraam: What are your thoughts on this?